<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate metric values — metric.values • BioMonTools</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate metric values — metric.values"><meta property="og:description" content="This function calculates metric values for bugs, fish, algae
, and coral. Inputs are a data frame with SampleID and taxa with phylogenetic
and autecological information (see below for required fields by community).
The dplyr package is used to generate the metric values."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BioMonTools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.2.9029</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/vignette_BioMonTools.html">Vignette, BioMonTools</a>
    </li>
    <li>
      <a href="../articles/vignette_MapTaxaObs.html">Taxa Maps</a>
    </li>
    <li>
      <a href="../articles/vignette_NewIndex.html">Vignette, Adding a New Index</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/leppott/BioMonTools/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate metric values</h1>
    <small class="dont-index">Source: <a href="https://github.com/leppott/BioMonTools/blob/HEAD/R/metric_values.R" class="external-link"><code>R/metric_values.R</code></a></small>
    <div class="hidden name"><code>metric.values.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function calculates metric values for bugs, fish, algae
, and coral. Inputs are a data frame with SampleID and taxa with phylogenetic
and autecological information (see below for required fields by community).
The dplyr package is used to generate the metric values.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">metric.values</span><span class="op">(</span></span>
<span>  <span class="va">fun.DF</span>,</span>
<span>  <span class="va">fun.Community</span>,</span>
<span>  fun.MetricNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  boo.Adjust <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fun.cols2keep <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  boo.marine <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  boo.Shiny <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  metric_subset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  taxaid_dni <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>fun.DF</dt>
<dd><p>Data frame of taxa (list required fields)</p></dd>


<dt>fun.Community</dt>
<dd><p>Community name for which to calculate metric values
(bugs, fish, algae, or coral)</p></dd>


<dt>fun.MetricNames</dt>
<dd><p>Optional vector of metric names to be returned.
If none are supplied then all will be returned.  Default=NULL</p></dd>


<dt>boo.Adjust</dt>
<dd><p>Optional boolean value on whether to perform adjustments of
values prior to scoring.  Default = FALSE but may be TRUE for certain
metrics.</p></dd>


<dt>fun.cols2keep</dt>
<dd><p>Column names of fun.DF to retain in the output.  Uses
column names.</p></dd>


<dt>boo.marine</dt>
<dd><p>Should estuary/marine metrics be included.
Ignored if fun.MetricNames is not null. Default = FALSE.</p></dd>


<dt>boo.Shiny</dt>
<dd><p>Boolean value for if the function is accessed via Shiny.
Default = FALSE.</p></dd>


<dt>verbose</dt>
<dd><p>Include messages to track progress.  Default = FALSE</p></dd>


<dt>metric_subset</dt>
<dd><p>Subset of metrics to be generated.  Internal function.
Default = NULL</p></dd>


<dt>taxaid_dni</dt>
<dd><p>Taxa names to be included in DNI (Do Not Include) metrics
(n = 3) but dropped for all other metrics.  Only for benthic metrics.
Default = NULL</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>data frame of SampleID and metric values</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>All percent metric results are 0-100.</p>
<p>No manipulations of the taxa are performed by this routine.
All benthic macroinvertebrate taxa should be identified to the appropriate
operational taxonomic unit (OTU).</p>
<p>Any non-count taxa should be identified in the "Exclude" field as "TRUE".
These taxa will be excluded from taxa richness metrics (but will count for
all others).</p>
<p>Any non-target taxa should be identified in the "NonTarget"
field as "TRUE".  Non-target taxa are those that are not part of your
intended #' capture list; e.g., fish,  herps, water column taxa, or water
surface taxa in a benthic sample.  The target list will vary by program.  The
non-target taxa will be removed prior to any calculations.</p>
<p>Excluded taxa are ambiguous taxa (on a sample basis), i.e.,
the parent taxa when child taxa are present.  For example, the parent taxa
Chironomidae would be excluded when the child taxa Tanytarsini is present.
Both would be excluded when Tanytarsus is present.  The markExcluded function
can be used to populated this field.</p>
<p>There are a number of required fields (see below) for metric to calculation.
If any fields are missing the user will be prompted as to which are missing
and if the user wants to continue or quit.  If the user continues the missing
fields will be added but will be filled with zero or NA (as appropriate).
Any metrics based on the missing fields will not be valid.</p>
<p>A future update may turn these fields into function parameters.  This would
allow the user to tweak the function inputs to match their data rather than
having to update their data to match the function.</p>
<p>Required fields, all communities:</p>
<p>* SAMPLEID (character or number, must be unique)</p>
<p>* TAXAID (character or number, must be unique)</p>
<p>* N_TAXA</p>
<p>* INDEX_NAME</p>
<p>* INDEX_CLASS (BCG or MMI site category; e.g., for BCG PacNW valid values
are "hi" or "lo")</p>
<p>Additional Required fields, bugs:</p>
<p>* EXCLUDE (valid values are TRUE and FALSE)</p>
<p>* NONTARGET (valid values are TRUE and FALSE)</p>
<p>* PHYLUM, SUBPHYLUM, CLASS, SUBCLASS, INFRAORDER, ORDER, FAMILY, SUBFAMILY,
TRIBE, GENUS</p>
<p>* FFG, HABIT, LIFE_CYCLE, TOLVAL, BCG_ATTR, THERMAL_INDICATOR, FFG2, TOLVAL2,
LONGLIVED, NOTEWORTHY, HABITAT, UFC, ELEVATION_ATTR, GRADIENT_ATTR,
WSAREA_ATTR, HABSTRUCT</p>
<p>Additional Required fields, fish:</p>
<p>* N_ANOMALIES</p>
<p>* SAMP_BIOMASS (biomass total for sample, funciton uses max in case entered
for all taxa in sample)</p>
<p>* DA_MI2, SAMP_WIDTH_M, SAMP_LENGTH_M, , TYPE, TOLER, NATIVE, TROPHIC, SILT,
FAMILY, GENUS, HYBRID, BCG_ATTR, THERMAL_INDICATOR, ELEVATION_ATTR,
GRADIENT_ATTR, WSAREA_ATTR, REPRODUCTION, HABITAT, CONNECTIVITY, SCC</p>
<p>Additional Required fields, algae:</p>
<p>* EXCLUDE, NONTARGET, PHYLUM, ORDER, FAMILY, GENUS, BC_USGS, TROPHIC_USGS,
SAP_USGS, PT_USGS, O_USGS, SALINITY_USGS, BAHLS_USGS, P_USGS, N_USGS,
HABITAT_USGS, N_FIXER_USGS, MOTILITY_USGS, SIZE_USGS, HABIT_USGS,
MOTILE2_USGS, TOLVAL, DIATOM_ISA, DIAT_CL, POLL_TOL, BEN_SES, DIATAS_TP,
DIATAS_TN, DIAT_COND, DIAT_CA, MOTILITY, NF</p>
<p>Valid values for fields:</p>
<p>* FFG: CG, CF, PR, SC, SH</p>
<p>* HABIT: BU, CB, CN, SP, SW</p>
<p>* LIFE_CYCLE: UNI, SEMI, MULTI</p>
<p>* THERMAL_INDICATOR: STENOC, COLD, COOL, WARM, STENOW, EURYTHERMAL
, COWA, NA</p>
<p>* LONGLIVED: TRUE, FALSE</p>
<p>* NOTEWORTHY: TRUE, FALSE</p>
<p>* HABITAT: BRAC, DEPO, GENE, HEAD, RHEO, RIVE, SPEC, UNKN</p>
<p>* UFC: integers 1:6 (taxonomic uncertainty frequency class)</p>
<p>* ELEVATION_ATTR: LOW, HIGH</p>
<p>* GRADIENT_ATTR: LOW, MOD, HIGH</p>
<p>* WSAREA_ATTR: SMALL, MEDIUM, LARGE, XLARGE</p>
<p>* REPRODUCTION: BROADCASTER, SIMPLE NEST, COMPLEX NEST, BEARER, MIGRATORY</p>
<p>* CONNECTIVITY: TRUE, FALSE</p>
<p>* SCC (Species of Conservation Concern): TRUE, FALSE</p>
<p>'Columns to keep' are additional fields in the input file that the user wants
retained in the output.  Fields need to be those that are unique per sample
and not associated with the taxa.  For example, the fields used in
qc.check(); Area_mi2, SurfaceArea, Density_m2, and Density_ft2.</p>
<p>If fun.MetricNames is provided only those metrics will be returned in the
provided order. This variable can be used to sort the metrics per the user's
preferences. By default the metric names will be returned in the groupings
that were used for calculation.</p>
<p>The fields TOLVAL2 and FFG2 are provided to allow the user to calculate
metrics based on alternative scenarios.  For example, including both HBI and
NCBI where the NCBI uses a different set of tolerance values (TOLVAL2).</p>
<p>If TAXAID is 'NONE' and N_TAXA is '0' then metrics **will** be calculated
with that record. Other values for TAXAID with N_TAXA = 0 will be removed
before calculations.</p>
<p>For 'Oligochete' metrics either Class or Subclass is required for
calculation.</p>
<p>The parameter boo.Shiny can be set to TRUE when accessing this function in
Shiny. Normally the QC check for required fields is interactive.  Setting
boo.Shiny to TRUE will always continue.  The default is FALSE.</p>
<p>The parameter 'taxaid_dni' denotes taxa to be included in Do Not Include
(DNI) metrics but dropped from all other metrics.  Only for benthic metrics.</p>
<p>Breaking change from 0.5 to 0.6 with change from Index_Name to Index_Class.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Example 1, data already in R</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">df_metric_values_bugs</span> <span class="op">&lt;-</span> <span class="fu">metric.values</span><span class="op">(</span><span class="va">data_benthos_PacNW</span>, <span class="st">"bugs"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(SAMPLEID, INDEX_NAME, INDEX_CLASS)`</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># View Results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html" class="external-link">View</a></span><span class="op">(</span><span class="va">df_metric_values_bugs</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span></span></span>
<span class="r-in"><span><span class="co"># Example 2, data from Excel</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># Packages</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org" class="external-link">readxl</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">df_samps_bugs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_excel</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/Data_Benthos.xlsx"</span></span></span>
<span class="r-in"><span>                                       , package <span class="op">=</span> <span class="st">"BioMonTools"</span><span class="op">)</span></span></span>
<span class="r-in"><span>                            , guess_max <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="fl">6</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Columns to keep</span></span></span>
<span class="r-in"><span><span class="va">myCols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Area_mi2"</span>, <span class="st">"SurfaceArea"</span>, <span class="st">"Density_m2"</span>, <span class="st">"Density_ft2"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run Function</span></span></span>
<span class="r-in"><span><span class="va">df_metric_values_bugs</span> <span class="op">&lt;-</span> <span class="fu">metric.values</span><span class="op">(</span><span class="va">df_samps_bugs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="op">]</span></span></span>
<span class="r-in"><span>                                       , <span class="st">"bugs"</span></span></span>
<span class="r-in"><span>                                       , fun.cols2keep <span class="op">=</span> <span class="va">myCols</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># View Results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html" class="external-link">View</a></span><span class="op">(</span><span class="va">df_metric_values_bugs</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get data in long format so can QC results more easily</span></span></span>
<span class="r-in"><span><span class="va">df_long</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">df_metric_values_bugs</span>, id.vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SAMPLEID"</span></span></span>
<span class="r-in"><span>                                                 , <span class="st">"INDEX_NAME"</span></span></span>
<span class="r-in"><span>                                                 , <span class="st">"INDEX_CLASS"</span></span></span>
<span class="r-in"><span>                                                 , <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">toupper</a></span><span class="op">(</span><span class="va">myCols</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>                          , variable.name <span class="op">=</span> <span class="st">"METRIC_NAME"</span></span></span>
<span class="r-in"><span>                          , value.name <span class="op">=</span> <span class="st">"METRIC_VALUE"</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in eval(expr, envir, enclos):</span> object 'myCols' not found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># Save Results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">df_long</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"metric.values.tsv"</span><span class="op">)</span></span></span>
<span class="r-in"><span>            , col.names <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># DataExplorer Report</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://boxuancui.github.io/DataExplorer/" class="external-link">DataExplorer</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="http://boxuancui.github.io/DataExplorer/reference/create_report.html" class="external-link">create_report</a></span><span class="op">(</span><span class="va">df_metric_values_bugs</span></span></span>
<span class="r-in"><span>              , output_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>                                 , <span class="st">"DataExplorer_Report_MetricValues.html"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="http://boxuancui.github.io/DataExplorer/reference/create_report.html" class="external-link">create_report</a></span><span class="op">(</span><span class="va">df_samps_bugs</span></span></span>
<span class="r-in"><span>              , output_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>                                   , <span class="st">"DataExplorer_Report_BugSamples.html"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span></span></span>
<span class="r-in"><span><span class="co"># Example 3, specific metrics or metrics in a specific order</span></span></span>
<span class="r-in"><span><span class="co">## reuse df_samps_bugs from above</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># metric names to keep (in this order)</span></span></span>
<span class="r-in"><span><span class="va">myMetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ni_total"</span>, <span class="st">"nt_EPT"</span>, <span class="st">"nt_Ephem"</span>, <span class="st">"pi_tv_intol"</span>, <span class="st">"pi_Ephem"</span></span></span>
<span class="r-in"><span>               , <span class="st">"nt_ffg_scrap"</span>, <span class="st">"pi_habit_climb"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run Function</span></span></span>
<span class="r-in"><span><span class="va">df_metric_values_bugs_myMetrics</span> <span class="op">&lt;-</span> <span class="fu">metric.values</span><span class="op">(</span><span class="va">df_samps_bugs</span>, <span class="st">"bugs"</span></span></span>
<span class="r-in"><span>                                               , fun.MetricNames <span class="op">=</span> <span class="va">myMetrics</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in eval(expr, envir, enclos):</span> object 'df_samps_bugs' not found</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># View Results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html" class="external-link">View</a></span><span class="op">(</span><span class="va">df_metric_values_bugs_myMetrics</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span></span></span>
<span class="r-in"><span><span class="co"># Example 4, fish metrics</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">df_metric_values_fish</span> <span class="op">&lt;-</span> <span class="fu">metric.values</span><span class="op">(</span><span class="va">data_fish_MBSS</span>, <span class="st">"fish"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> There are 1 missing fields in the data:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> BCG_ATTR2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> If you continue the metrics associated with these fields will be invalid.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> For example, if the NATIVE field is missing all native related metrics will not be correct.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Do you wish to continue (YES or NO)?</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> boo.Shiny == TRUE and interactive == FALSE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>               so prompt skipped and value set to '1'.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Metrics related to the following fields are invalid:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>    BCG_ATTR2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> EXCLUDE column does not have any TRUE values.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> This is common with fish samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Valid values are TRUE or FALSE.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Other values are not recognized</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in dplyr::summarise(dplyr::group_by(myDF, SAMPLEID, INDEX_NAME,     INDEX_CLASS, SAMP_WIDTH_M, SAMP_LENGTH_M), .groups = "drop_last",     ni_total = sum(N_TAXA, na.rm = TRUE), ni_total_notoler = sum(N_TAXA[TOLER !=         "TOLERANT" | is.na(TOLER)], na.rm = TRUE), ni_natnonhybridnonmf = sum(N_TAXA[NATIVE ==         "NATIVE" &amp; (HYBRID != TRUE | is.na(HYBRID)) &amp; (TYPE !=         "MOSQUITOFISH" | is.na(TYPE))], na.rm = TRUE), ni_natnonhybridnonmfnonLepomis = sum(N_TAXA[NATIVE ==         "NATIVE" &amp; (HYBRID != TRUE | is.na(HYBRID)) &amp; (TYPE !=         "MOSQUITOFISH" | is.na(TYPE)) &amp; (GENUS != "LEPOMIS" |         is.na(GENUS))], na.rm = TRUE), pi_AmmEthPerc = 100 *         sum(N_TAXA[GENUS == "AMMOCRYPTA" | GENUS == "ETHEOSTOMA" |             GENUS == "PERCINA"], na.rm = TRUE)/ni_total, pi_AmmEthPerc_Cott_Notur = 100 *         sum(N_TAXA[(GENUS == "AMMOCRYPTA" | GENUS == "ETHEOSTOMA" |             GENUS == "PERCINA" | GENUS == "NOTURUS") | FAMILY ==             "COTTIDAE"], na.rm = TRUE)/ni_total, pi_rbs = 100 *         sum(N_TAXA[TYPE == "RBS"], na.rm = TRUE)/ni_total, pi_brooktrout = 100 *         sum(N_TAXA[TYPE == "BROOK TROUT"], na.rm = TRUE)/ni_total,     pi_brooktrout_wild = 100 * sum(N_TAXA[TAXAID == "BROOK TROUT, WILD"],         na.rm = TRUE)/ni_total, pi_Cato = 100 * sum(N_TAXA[FAMILY ==         "CATOSTOMIDAE"], na.rm = TRUE)/ni_total, pi_Cent = 100 *         sum(N_TAXA[FAMILY == "CENTRARCHIDAE"], na.rm = TRUE)/ni_total,     pi_natCent = 100 * sum(N_TAXA[NATIVE == "NATIVE" &amp; FAMILY ==         "CENTRARCHIDAE"], na.rm = TRUE)/ni_total, pi_Cott = 100 *         sum(N_TAXA[FAMILY == "COTTIDAE"], na.rm = TRUE)/ni_total,     pi_Cyprin = 100 * sum(N_TAXA[FAMILY == "CYPRINIDAE"], na.rm = TRUE)/ni_total,     pi_Ictal = 100 * sum(N_TAXA[FAMILY == "ICTALURIDAE"], na.rm = TRUE)/ni_total,     pi_native = 100 * sum(N_TAXA[NATIVE == "NATIVE"], na.rm = TRUE)/ni_total,     pi_nonnative = 100 * sum(N_TAXA[is.na(NATIVE) | NATIVE !=         "NATIVE"], na.rm = TRUE)/ni_total, pi_Notur = 100 * sum(N_TAXA[GENUS ==         "NOTURUS"], na.rm = TRUE)/ni_total, pi_sculpin = 100 *         sum(N_TAXA[TYPE == "SCULPIN"], na.rm = TRUE)/ni_total,     pi_Lepomis = 100 * sum(N_TAXA[GENUS == "LEPOMIS"], na.rm = TRUE)/ni_total,     pi_Salm = 100 * sum(N_TAXA[FAMILY == "SALMONIDAE"], na.rm = TRUE)/ni_total,     pi_trout = 100 * sum(N_TAXA["TROUT" %in% TYPE], na.rm = TRUE)/ni_total,     pi_brooktrout_BCG_att6 = 100 * (sum(N_TAXA[TYPE == "BROOK TROUT"],         na.rm = TRUE) + sum(N_TAXA[BCG_ATTR == "6"], na.rm = TRUE))/ni_total,     pi_connect = 100 * sum(N_TAXA[CONNECTIVITY == TRUE], na.rm = TRUE)/ni_total,     pi_scc = 100 * sum(N_TAXA[SCC == TRUE], na.rm = TRUE)/ni_total,     pi_brooktrout2brooktrout_BCG_att6 = 100 * sum(N_TAXA[TYPE ==         "BROOK TROUT"], na.rm = TRUE)/(sum(N_TAXA[TYPE == "BROOK TROUT"],         na.rm = TRUE) + sum(N_TAXA[BCG_ATTR == "6"], na.rm = TRUE)),     pi_bfs = 100 * sum(N_TAXA[(TYPE == "BENTHIC" &amp; TROPHIC_IV ==         TRUE) | TYPE == "RBS" | TYPE == "SMM"], na.rm = TRUE)/ni_total,     nt_total = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp; N_TAXA &gt;         0], na.rm = TRUE), nt_benthic = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TYPE == "BENTHIC"], na.rm = TRUE), nt_AmmEthPerc = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (GENUS == "AMMOCRYPTA" | GENUS == "ETHEOSTOMA" |         GENUS == "PERCINA")], na.rm = TRUE), nt_AmmEthPerc_Cott_Notur = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (GENUS == "AMMOCRYPTA" | GENUS == "ETHEOSTOMA" |         GENUS == "PERCINA" | GENUS == "NOTURUS") | FAMILY ==         "COTTIDAE"], na.rm = TRUE), nt_Cato = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; FAMILY == "CATOSTOMIDAE"], na.rm = TRUE), nt_Cent = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; FAMILY == "CENTRARCHIDAE"], na.rm = TRUE), nt_natCent = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; NATIVE == "NATIVE" &amp; FAMILY == "CENTRARCHIDAE"],         na.rm = TRUE), nt_Cott = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; FAMILY == "COTTIDAE"], na.rm = TRUE), nt_Cyprin = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; FAMILY == "CYPRINIDAE"], na.rm = TRUE), nt_natCyprin = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; NATIVE == "NATIVE" &amp; FAMILY == "CYPRINIDAE"],         na.rm = TRUE), nt_Lepomis = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; GENUS == "LEPOMIS"], na.rm = TRUE), nt_native = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; NATIVE == "NATIVE" &amp; N_TAXA &gt; 0], na.rm = TRUE),     nt_nonnative = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         (is.na(NATIVE) | NATIVE != "NATIVE") &amp; N_TAXA &gt; 0], na.rm = TRUE),     nt_nativenonhybrid = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; NATIVE == "NATIVE" &amp; (HYBRID != TRUE | is.na(HYBRID))],         na.rm = TRUE), nt_Notur = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; GENUS == "NOTURUS"], na.rm = TRUE), nt_Ictal = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; FAMILY == "ICTALURIDAE"], na.rm = TRUE), nt_natsunfish = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; NATIVE == "NATIVE" &amp; TYPE == "SUNFISH"], na.rm = TRUE),     nt_natCent_sunfish = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; NATIVE == "NATIVE" &amp; (TYPE == "SUNFISH" | TYPE ==         "CENTRARCHIDAE")], na.rm = TRUE), nt_natCent = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; NATIVE == "NATIVE" &amp; FAMILY == "CENTRARCHIDAE"],         na.rm = TRUE), nt_natinsctCypr = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; NATIVE == "NATIVE" &amp; TROPHIC_IS == TRUE &amp; FAMILY ==         "CYPRINIDAE"], na.rm = TRUE), nt_natrbs = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; NATIVE == "NATIVE" &amp; TYPE == "RBS"], na.rm = TRUE),     nt_Petro = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp; FAMILY ==         "PETROMYZONTIDAE"], na.rm = TRUE), nt_Salm = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; FAMILY == "SALMONIDAE"], na.rm = TRUE), nt_connect = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; CONNECTIVITY == TRUE], na.rm = TRUE), nt_scc = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; SCC == TRUE], na.rm = TRUE), nt_beninsct_nows_nobg = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_IS == TRUE &amp; (TAXAID != "CATOSTOMUS COMMERSONII" |         TAXAID != "LEPOMIS MACROCHIRUS" | is.na(TAXAID))], na.rm = TRUE),     nt_trout_sunfish_notoler = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; ("TROUT" %in% TYPE | TYPE == "SUNFISH") &amp; (TOLER !=         "TOLERANT" | is.na(TOLER))], na.rm = TRUE), pt_AmmEthPerc = 100 *         nt_AmmEthPerc/nt_total, pt_AmmEthPerc_Cott_Notur = 100 *         nt_AmmEthPerc_Cott_Notur/nt_total, pt_Cato = 100 * nt_Cato/nt_total,     pt_Cent = 100 * nt_Cent/nt_total, pt_natCent = 100 * nt_natCent/nt_total,     pt_Cott = 100 * nt_Cott/nt_total, pt_Cyprin = 100 * nt_Cyprin/nt_total,     pt_Ictal = 100 * nt_Ictal/nt_total, pt_Lepomis = 100 * nt_Lepomis/nt_total,     pt_native = 100 * nt_native/nt_total, pt_nonnative = 100 *         nt_nonnative/nt_total, pt_Notur = 100 * nt_Notur/nt_total,     pt_Salm = 100 * nt_Salm/nt_total, pt_connect = 100 * nt_connect/nt_total,     pt_scc = 100 * nt_scc/nt_total, nt_beninvert = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TYPE == "BENTHIC" &amp; TROPHIC_IV == TRUE], na.rm = TRUE),     nt_habitat_beninvert = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_IV == TRUE &amp; HABITAT_B == TRUE], na.rm = TRUE),     nt_detritivore = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         TROPHIC_DE == TRUE], na.rm = TRUE), nt_herbivore = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_HB == TRUE], na.rm = TRUE), nt_omnivore = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_OM == TRUE], na.rm = TRUE), nt_planktivore = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_PL == TRUE], na.rm = TRUE), nt_topcarn = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_TC == TRUE], na.rm = TRUE), nt_piscivore = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_PI == TRUE], na.rm = TRUE), pi_lithophil = 100 *         sum(N_TAXA[SILT == TRUE], na.rm = TRUE)/ni_total, pi_habitat_beninvert = 100 *         sum(N_TAXA[TROPHIC_IV == TRUE &amp; HABITAT_B == TRUE], na.rm = TRUE)/ni_total,     pi_detritivore = 100 * sum(N_TAXA[TROPHIC_DE == TRUE], na.rm = TRUE)/ni_total,     pi_genomninvrt = 100 * sum(N_TAXA[TROPHIC_GE == TRUE | TROPHIC_OM ==         TRUE | TROPHIC_IV == TRUE], na.rm = TRUE)/ni_total, pi_herbivore = 100 *         sum(N_TAXA[TROPHIC_HB == TRUE], na.rm = TRUE)/ni_total,     pi_insectivore = 100 * sum(N_TAXA[TROPHIC_IS == TRUE], na.rm = TRUE)/ni_total,     pi_insctCypr = 100 * sum(N_TAXA[TROPHIC_IS == TRUE &amp; FAMILY ==         "CYPRINIDAE"], na.rm = TRUE)/ni_total, pi_gen = 100 *         sum(N_TAXA[TROPHIC_GE == TRUE], na.rm = TRUE)/ni_total,     pi_genherb = 100 * sum(N_TAXA[TROPHIC_GE == TRUE | TROPHIC_HB ==         TRUE], na.rm = TRUE)/ni_total, pi_omnivore = 100 * sum(N_TAXA[TROPHIC_OM ==         TRUE], na.rm = TRUE)/ni_total, pi_planktivore = 100 *         sum(N_TAXA[TROPHIC_PL == TRUE], na.rm = TRUE)/ni_total,     pi_topcarn = 100 * sum(N_TAXA[TROPHIC_TC == TRUE], na.rm = TRUE)/ni_total,     pi_trout = 100 * sum(N_TAXA["TROUT" %in% TYPE], na.rm = TRUE)/ni_total,     pi_pisc_noae = 100 * sum(N_TAXA[TYPE == "PISCIVORE" &amp; (TAXAID !=         "ANGUILLA ROSTRATA" | is.na(TAXAID))], na.rm = TRUE)/ni_total,     pt_habitat_beninvert = 100 * nt_habitat_beninvert/nt_total,     pt_detritivore = 100 * nt_detritivore/nt_total, pt_herbivore = 100 *         nt_herbivore/nt_total, pt_omnivore = 100 * nt_omnivore/nt_total,     pt_planktivore = 100 * nt_planktivore/nt_total, pt_topcarn = 100 *         nt_topcarn/nt_total, nt_tv_intol = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TOLER == "INTOLERANT"], na.rm = TRUE), nt_tv_intolhwi = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (TOLER == "INTOLERANT" | TOLER == "HWI")], na.rm = TRUE),     pi_tv_toler = 100 * sum(N_TAXA[TOLER == "TOLERANT"], na.rm = TRUE)/ni_total,     x_Shan_e = -sum((N_TAXA/ni_total) * log((N_TAXA/ni_total)),         na.rm = TRUE), x_Shan_2 = x_Shan_e/log(2), x_Shan_10 = x_Shan_e/log(10),     x_Evenness = x_Shan_e/log(nt_total), x_Evenness100_ni99gt = ifelse(ni_total &lt;         100, 1, x_Evenness * 100), length_m = max(SAMP_LENGTH_M,         na.rm = TRUE), area_m2 = max(SAMP_WIDTH_M, na.rm = TRUE) *         length_m, ni_m2 = ni_total/area_m2, ni_200m = 200 * ni_total/length_m,     ni_natnonhybridnonmf_200m = 200 * ni_natnonhybridnonmf/length_m,     ni_natnonhybridnonmfnonLepomis_200m = 200 * ni_natnonhybridnonmfnonLepomis/length_m,     biomass_m2 = max(SAMP_BIOMASS, na.rm = TRUE)/area_m2, pi_anomalies = 100 *         sum(N_ANOMALIES, na.rm = TRUE)/ni_total, pi_delt = 100 *         sum(N_ANOMALIES, na.rm = TRUE)/ni_total, pi_dom01 = 100 *         max(N_TAXA, na.rm = TRUE)/ni_total, pi_dom02 = 100 *         max(ni_dom02, na.rm = TRUE)/ni_total, pi_dom03 = 100 *         max(ni_dom03, na.rm = TRUE)/ni_total, pi_dom04 = 100 *         max(ni_dom04, na.rm = TRUE)/ni_total, pi_dom05 = 100 *         max(ni_dom05, na.rm = TRUE)/ni_total, pi_dom06 = 100 *         max(ni_dom06, na.rm = TRUE)/ni_total, pi_dom07 = 100 *         max(ni_dom07, na.rm = TRUE)/ni_total, pi_dom08 = 100 *         max(ni_dom08, na.rm = TRUE)/ni_total, pi_dom09 = 100 *         max(ni_dom09, na.rm = TRUE)/ni_total, pi_dom10 = 100 *         max(ni_dom10, na.rm = TRUE)/ni_total, nt_BCG_att1 = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; BCG_ATTR == "1"], na.rm = TRUE), nt_BCG_att12 = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (BCG_ATTR == "1" | BCG_ATTR == "2")], na.rm = TRUE),     nt_BCG_att123 = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         (BCG_ATTR == "1" | BCG_ATTR == "2" | BCG_ATTR == "3")],         na.rm = TRUE), nt_BCG2_att123b = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (BCG_ATTR == "1" | BCG_ATTR == "2" | BCG_ATTR2 ==         "3_BETTER")], na.rm = TRUE), nt_BCG_att1234 = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (BCG_ATTR == "1" | BCG_ATTR == "2" | BCG_ATTR ==         "3" | BCG_ATTR == "4")], na.rm = TRUE), nt_BCG2_att1234b = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (BCG_ATTR == "1" | BCG_ATTR == "2" | BCG_ATTR ==         "3" | BCG_ATTR2 == "4_BETTER")], na.rm = TRUE), nt_BCG_att1236 = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (BCG_ATTR == "1" | BCG_ATTR == "2" | BCG_ATTR ==         "3" | BCG_ATTR == "6")], na.rm = TRUE), nt_BCG_att1236b = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (BCG_ATTR == "1" | BCG_ATTR == "2" | BCG_ATTR ==         "3" | BCG_ATTR == "6B")], na.rm = TRUE), nt_BCG_att12346 = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (BCG_ATTR == "1" | BCG_ATTR == "2" | BCG_ATTR ==         "3" | BCG_ATTR == "4" | BCG_ATTR == "6")], na.rm = TRUE),     nt_BCG_att12346b = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         (BCG_ATTR == "1" | BCG_ATTR == "2" | BCG_ATTR == "3" |             BCG_ATTR == "4" | BCG_ATTR == "6B")], na.rm = TRUE),     nt_BCG_att1i236i = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         (BCG_ATTR == "1I" | BCG_ATTR == "2" | BCG_ATTR == "3" |             BCG_ATTR == "6I")], na.rm = TRUE), nt_BCG_att2 = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; BCG_ATTR == "2"], na.rm = TRUE), nt_BCG_att2native = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; BCG_ATTR == "2" &amp; NATIVE == "NATIVE"], na.rm = TRUE),     nt_BCG_att23_scc = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         (BCG_ATTR == "2" | BCG_ATTR == "3") &amp; SCC == TRUE], na.rm = TRUE),     nt_BCG_att3 = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         BCG_ATTR == "3"], na.rm = TRUE), nt_BCG_att3native = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; BCG_ATTR == "3" &amp; NATIVE == "NATIVE"], na.rm = TRUE),     nt_BCG_att4 = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         BCG_ATTR == "4"], na.rm = TRUE), nt_BCG_att4native = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; BCG_ATTR == "4" &amp; NATIVE == "NATIVE"], na.rm = TRUE),     nt_BCG_att5 = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         BCG_ATTR == "5"], na.rm = TRUE), nt_BCG_att5native = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; BCG_ATTR == "5" &amp; NATIVE == "NATIVE"], na.rm = TRUE),     nt_BCG_att55a6 = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         (BCG_ATTR == "5" | BCG_ATTR == "5A" | BCG_ATTR == "6")],         na.rm = TRUE), nt_BCG_att55a6a = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (BCG_ATTR == "5" | BCG_ATTR == "5A" | BCG_ATTR ==         "6A")], na.rm = TRUE), nt_BCG_att56t = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (BCG_ATTR == "5" | BCG_ATTR == "6T")], na.rm = TRUE),     nt_BCG_att6i = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         BCG_ATTR == "6I"], na.rm = TRUE), nt_BCG_att6m = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; BCG_ATTR == "6M"], na.rm = TRUE), nt_BCG_att6t = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; BCG_ATTR == "6T"], na.rm = TRUE), nt_BCG_attNA = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; is.na(BCG_ATTR)], na.rm = TRUE), pi_BCG_att12 = 100 *         sum(N_TAXA[(BCG_ATTR == "1" | BCG_ATTR == "2")], na.rm = TRUE)/ni_total,     pi_BCG_att123 = 100 * sum(N_TAXA[(BCG_ATTR == "1" | BCG_ATTR ==         "2" | BCG_ATTR == "3")], na.rm = TRUE)/ni_total, pi_BCG_att1234 = 100 *         sum(N_TAXA[(BCG_ATTR == "1" | BCG_ATTR == "2" | BCG_ATTR ==             "3" | BCG_ATTR == "4")], na.rm = TRUE)/ni_total,     pi_BCG_att1236 = 100 * sum(N_TAXA[(BCG_ATTR == "1" | BCG_ATTR ==         "2" | BCG_ATTR == "3" | BCG_ATTR == "6")], na.rm = TRUE)/ni_total,     pi_BCG_att1236sp = 100 * sum(N_TAXA[(BCG_ATTR == "1" | BCG_ATTR ==         "2" | BCG_ATTR == "3" | BCG_ATTR == "6")], na.rm = TRUE)/sum(N_TAXA[(BCG_ATTR ==         "1" | BCG_ATTR == "2" | BCG_ATTR == "3" | BCG_ATTR ==         "6" | BCG_ATTR == "5" | BCG_ATTR == "5A" | BCG_ATTR ==         "6A")], na.rm = TRUE), pi_BCG_att1236b = 100 * sum(N_TAXA[(BCG_ATTR ==         "1" | BCG_ATTR == "2" | BCG_ATTR == "3" | BCG_ATTR ==         "6B")], na.rm = TRUE)/ni_total, pi_BCG_att12346b = 100 *         sum(N_TAXA[(BCG_ATTR == "1" | BCG_ATTR == "2" | BCG_ATTR ==             "3" | BCG_ATTR == "4" | BCG_ATTR == "6B")], na.rm = TRUE)/ni_total,     pi_BCG_att1i236i = 100 * sum(N_TAXA[(BCG_ATTR == "1I" | BCG_ATTR ==         "2" | BCG_ATTR == "3" | BCG_ATTR == "6I")], na.rm = TRUE)/ni_total,     pi_BCG_att2 = 100 * sum(N_TAXA[BCG_ATTR == "2"], na.rm = TRUE)/ni_total,     pi_BCG_att2native = 100 * sum(N_TAXA[BCG_ATTR == "2" &amp; NATIVE ==         "NATIVE"], na.rm = TRUE)/ni_total, pi_BCG_att23_scc = 100 *         sum(N_TAXA[(BCG_ATTR == "2" | BCG_ATTR == "3") &amp; SCC ==             TRUE], na.rm = TRUE)/ni_total, pi_BCG_att3 = 100 *         sum(N_TAXA[BCG_ATTR == "3"], na.rm = TRUE)/ni_total,     pi_BCG_att3native = 100 * sum(N_TAXA[BCG_ATTR == "3" &amp; NATIVE ==         "NATIVE"], na.rm = TRUE)/ni_total, pi_BCG_att4 = 100 *         sum(N_TAXA[BCG_ATTR == "4"], na.rm = TRUE)/ni_total,     pi_BCG_att4native = 100 * sum(N_TAXA[BCG_ATTR == "4" &amp; NATIVE ==         "NATIVE"], na.rm = TRUE)/ni_total, pi_BCG_att5 = 100 *         sum(N_TAXA[BCG_ATTR == "5"], na.rm = TRUE)/ni_total,     pi_BCG_att5native = 100 * sum(N_TAXA[BCG_ATTR == "5" &amp; NATIVE ==         "NATIVE"], na.rm = TRUE)/ni_total, pi_BCG_att55a6a = 100 *         sum(N_TAXA[(BCG_ATTR == "5" | BCG_ATTR == "5A" | BCG_ATTR ==             "6A")], na.rm = TRUE)/ni_total, pi_BCG_att56a = 100 *         sum(N_TAXA[(BCG_ATTR == "5" | BCG_ATTR == "6A")], na.rm = TRUE)/ni_total,     pi_BCG_att5a6a = 100 * sum(N_TAXA[(BCG_ATTR == "5A" | BCG_ATTR ==         "6A")], na.rm = TRUE)/ni_total, pi_BCG_att56t = 100 *         sum(N_TAXA[(BCG_ATTR == "5" | BCG_ATTR == "6T")], na.rm = TRUE)/ni_total,     pi_BCG_att6 = 100 * sum(N_TAXA[BCG_ATTR == "6"], na.rm = TRUE)/ni_total,     pi_BCG_att6i = 100 * sum(N_TAXA[BCG_ATTR == "6I"], na.rm = TRUE)/ni_total,     pi_BCG_att6m = 100 * sum(N_TAXA[BCG_ATTR == "6M"], na.rm = TRUE)/ni_total,     pi_BCG_att6t = 100 * sum(N_TAXA[BCG_ATTR == "6T"], na.rm = TRUE)/ni_total,     pi_BCG_att66a = 100 * sum(N_TAXA[(BCG_ATTR == "6" | BCG_ATTR ==         "6A")], na.rm = TRUE)/ni_total, pi_BCG_att66a6b = 100 *         sum(N_TAXA[(BCG_ATTR == "6" | BCG_ATTR == "6A" | BCG_ATTR ==             "6B")], na.rm = TRUE)/ni_total, pi_BCG_att66s6t = 100 *         sum(N_TAXA[BCG_ATTR == "6" | BCG_ATTR == "6S" | BCG_ATTR ==             "6T"], na.rm = TRUE)/ni_total, pi_BCG_attNA = 100 *         sum(N_TAXA[is.na(BCG_ATTR)], na.rm = TRUE)/ni_total,     pt_BCG_att12 = 100 * nt_BCG_att12/nt_total, pt_BCG_att123 = 100 *         nt_BCG_att123/nt_total, pt_BCG2_att123b = 100 * nt_BCG2_att123b/nt_total,     pt_BCG_att1234 = 100 * nt_BCG_att1234/nt_total, pt_BCG2_att1234b = 100 *         nt_BCG2_att1234b/nt_total, pt_BCG_att1236 = 100 * nt_BCG_att1236/nt_total,     pt_BCG_att1236b = 100 * nt_BCG_att1236b/nt_total, pt_BCG_att1236sp = 100 *         nt_BCG_att1236/(nt_BCG_att1236 + nt_BCG_att55a6a), pt_BCG_att12346b = 100 *         nt_BCG_att12346b/nt_total, pt_BCG_att1i236i = 100 * nt_BCG_att1i236i/nt_total,     pt_BCG_att2 = 100 * nt_BCG_att2/nt_total, pt_BCG_att2native = 100 *         nt_BCG_att2native/nt_total, pt_BCG_att23_scc = 100 *         nt_BCG_att23_scc/nt_total, pt_BCG_att3 = 100 * nt_BCG_att3/nt_total,     pt_BCG_att3native = 100 * nt_BCG_att3native/nt_total, pt_BCG_att4 = 100 *         nt_BCG_att4/nt_total, pt_BCG_att4native = 100 * nt_BCG_att4native/nt_total,     pt_BCG_att5 = 100 * nt_BCG_att5/nt_total, pt_BCG_att5native = 100 *         nt_BCG_att5native/nt_total, pt_BCG_att55a6a = 100 * nt_BCG_att55a6/nt_total,     pt_BCG_att56t = 100 * nt_BCG_att56t/nt_total, pt_BCG_att6i = 100 *         nt_BCG_att6i/nt_total, pt_BCG_att6m = 100 * nt_BCG_att6m/nt_total,     pt_BCG_att6t = 100 * nt_BCG_att6t/nt_total, pt_BCG_attNA = 100 *         nt_BCG_attNA/nt_total, pi_dom01_BCG_att4 = 100 * max(0,         N_TAXA[(BCG_ATTR == "4")], na.rm = TRUE)/ni_total, pi_dom01_BCG_att45 = 100 *         max(0, N_TAXA[(BCG_ATTR == "4" | BCG_ATTR == "5")], na.rm = TRUE)/ni_total,     pi_dom01_BCG_att5 = 100 * max(0, N_TAXA[(BCG_ATTR == "5")],         na.rm = TRUE)/ni_total, pi_dom01_BCG_att5a = 100 * max(0,         N_TAXA[(BCG_ATTR == "5A")], na.rm = TRUE)/ni_total, pi_dom01_BCG_att5a6a = 100 *         max(0, N_TAXA[(BCG_ATTR == "5A" | BCG_ATTR == "6A")],             na.rm = TRUE)/ni_total, pi_dom01_BCG_att566a = 100 *         max(0, N_TAXA[(BCG_ATTR == "5" | BCG_ATTR == "6" | BCG_ATTR ==             "6A")], na.rm = TRUE)/ni_total, nt_ti_corecold = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TI_CORECOLD == TRUE], na.rm = TRUE), nt_ti_cold = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TI_COLD == TRUE], na.rm = TRUE), nt_ti_cool = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TI_COOL == TRUE], na.rm = TRUE), nt_ti_warm = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TI_WARM == TRUE], na.rm = TRUE), nt_ti_eury = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TI_EURY == TRUE], na.rm = TRUE), nt_ti_na = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TI_NA == TRUE], na.rm = TRUE), nt_ti_corecold_cold = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; (TI_CORECOLD == TRUE | TI_COLD == TRUE)], na.rm = TRUE),     nt_ti_cool_warm = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         (TI_COOL == TRUE | TI_WARM == TRUE)], na.rm = TRUE),     pi_ti_corecold = 100 * sum(N_TAXA[TI_CORECOLD == TRUE], na.rm = TRUE)/ni_total,     pi_ti_cold = 100 * sum(N_TAXA[TI_COLD == TRUE], na.rm = TRUE)/ni_total,     pi_ti_cool = 100 * sum(N_TAXA[TI_COOL == TRUE], na.rm = TRUE)/ni_total,     pi_ti_warm = 100 * sum(N_TAXA[TI_WARM == TRUE], na.rm = TRUE)/ni_total,     pi_ti_eury = 100 * sum(N_TAXA[TI_EURY == TRUE], na.rm = TRUE)/ni_total,     pi_ti_na = 100 * sum(N_TAXA[TI_NA == TRUE], na.rm = TRUE)/ni_total,     pi_ti_corecold_cold = 100 * sum(N_TAXA[TI_CORECOLD == TRUE |         TI_COLD == TRUE], na.rm = TRUE)/ni_total, pi_ti_cool_warm = 100 *         sum(N_TAXA[TI_COOL == TRUE | TI_WARM == TRUE], na.rm = TRUE)/ni_total,     pt_ti_corecold = 100 * nt_ti_corecold/nt_total, pt_ti_cold = 100 *         nt_ti_cold/nt_total, pt_ti_cool = 100 * nt_ti_cool/nt_total,     pt_ti_warm = 100 * nt_ti_warm/nt_total, pt_ti_eury = 100 *         nt_ti_eury/nt_total, pt_ti_na = 100 * nt_ti_na/nt_total,     pt_ti_corecold_cold = 100 * nt_ti_corecold_cold/nt_total,     pt_ti_cool_warm = 100 * nt_ti_cool_warm/nt_total, nt_elev_low = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; ELEVATION_LOW == TRUE], na.rm = TRUE), nt_elev_high = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; ELEVATION_HIGH == TRUE], na.rm = TRUE), nt_grad_low = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; GRADIENT_LOW == TRUE], na.rm = TRUE), nt_grad_mod = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; GRADIENT_MOD == TRUE], na.rm = TRUE), nt_grad_high = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; GRADIENT_HIGH == TRUE], na.rm = TRUE), nt_wsarea_small = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; WSAREA_S == TRUE], na.rm = TRUE), nt_wsarea_medium = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; WSAREA_M == TRUE], na.rm = TRUE), nt_wsarea_large = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; WSAREA_L == TRUE], na.rm = TRUE), nt_wsarea_xlarge = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; WSAREA_XL == TRUE], na.rm = TRUE), nt_repro_broadcaster = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; REPRO_BCAST == TRUE], na.rm = TRUE), nt_repro_nestsimp = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; REPRO_NS == TRUE], na.rm = TRUE), nt_repro_nestcomp = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; REPRO_NC == TRUE], na.rm = TRUE), nt_repro_bearer = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; REPRO_BEAR == TRUE], na.rm = TRUE), nt_repro_migratory = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; REPRO_MIG == TRUE], na.rm = TRUE), nt_repro_lithophil = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; REPRO_LITH == TRUE], na.rm = TRUE), pi_repro_lithophil = 100 *         sum(N_TAXA[REPRO_LITH == TRUE], na.rm = TRUE)/ni_total,     pt_repro_lithophil = 100 * nt_repro_lithophil/nt_total, nt_habitat_b = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; HABITAT_B == TRUE], na.rm = TRUE), nt_habitat_w = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; HABITAT_W == TRUE], na.rm = TRUE), nt_habitat_f = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; HABITAT_F == TRUE], na.rm = TRUE), pi_habitat_b = 100 *         sum(N_TAXA[HABITAT_B == TRUE], na.rm = TRUE)/ni_total,     pi_habitat_w = 100 * sum(N_TAXA[HABITAT_W == TRUE], na.rm = TRUE)/ni_total,     pi_habitat_f = 100 * sum(N_TAXA[HABITAT_F == TRUE], na.rm = TRUE)/ni_total,     pt_habitat_b = 100 * nt_habitat_b/nt_total, pt_habitat_w = 100 *         nt_habitat_w/nt_total, pt_habitat_f = 100 * nt_habitat_f/nt_total,     nt_piscivore_BCG_att66s6t = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_PI == TRUE &amp; (BCG_ATTR == "6" | BCG_ATTR ==         "6S" | BCG_ATTR == "6T")], na.rm = TRUE), nt_LLNLB = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TYPE == "LLNLB"], na.rm = TRUE), nt_Cyprin_BCG_att1234 = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; FAMILY == "CYPRINIDAE" &amp; (BCG_ATTR == "1" | BCG_ATTR ==         "2" | BCG_ATTR == "3" | BCG_ATTR == "4")], na.rm = TRUE),     ni_Hybognathus_amarus = sum(N_TAXA[TAXAID == "HYBOGNATHUS AMARUS"],         na.rm = TRUE), x_TrophicCats = dplyr::n_distinct(TROPHIC,         na.rm = TRUE), x_BCG_Mean = mean(TOLVAL2, na.rm = TRUE),     nt_PupKilli = dplyr::n_distinct(TAXAID[EXCLUDE != TRUE &amp;         (TAXAID == "CYPRINODON RUBROFLUVIATILIS" | TAXAID ==             "FUNDULUS KANSAE" | TAXAID == "FUNDULUS ZEBRINUS")],         na.rm = TRUE), ni_total_ExclSchool = sum(N_TAXA[TYPE_SCHOOL !=         TRUE], na.rm = TRUE), ni_total_notoler_mn = sum(N_TAXA[TOLER_T !=         TRUE], na.rm = TRUE), nt_coldwater = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; HABITAT_CW == TRUE], na.rm = TRUE), nt_natcoldwater = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; HABITAT_CWN == TRUE], na.rm = TRUE), nt_hw_notoler = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; HABITAT_HW_noT == TRUE], na.rm = TRUE), nt_serialspawner = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; REPRO_SER == TRUE], na.rm = TRUE), nt_simplelithophil = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; REPRO_SILI == TRUE], na.rm = TRUE), nt_tv_sens = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TOLER_S == TRUE], na.rm = TRUE), nt_tv_senscoldwater = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TOLER_SCW == TRUE], na.rm = TRUE), nt_tv_tolercoldwater = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TOLER_TCW == TRUE], na.rm = TRUE), nt_tv_toler = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TOLER_T == TRUE], na.rm = TRUE), nt_tv_vtoler = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TOLER_VT == TRUE], na.rm = TRUE), nt_beninsct_notoler = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_BI_noT == TRUE], na.rm = TRUE), nt_detritivore = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_DE == TRUE], na.rm = TRUE), nt_gen = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_GE == TRUE], na.rm = TRUE), nt_insectivore_notoler = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_IN_noT == TRUE], na.rm = TRUE), nt_omnivore = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TROPHIC_OM == TRUE], na.rm = TRUE), nt_dartersculpin = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TYPE_DS == TRUE], na.rm = TRUE), nt_darterscultpinsucker = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TYPE_DSS == TRUE], na.rm = TRUE), nt_pioneer = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TYPE_PI == TRUE], na.rm = TRUE), nt_shortlived = dplyr::n_distinct(TAXAID[EXCLUDE !=         TRUE &amp; TYPE_SL == TRUE], na.rm = TRUE), pi_hw_notoler_ExclSchool = 100 *         sum(N_TAXA[HABITAT_HW_noT == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_wetland_notoler_ExclSchool = 100 *         sum(N_TAXA[HABITAT_WE_noT == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_natcoldwater_ExclSchool = 100 *         sum(N_TAXA[HABITAT_CWN == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_ma2_ExclShool = 100 *         sum(N_TAXA[REPRO_MA2 == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_ma3_notoler_ExclSchool = 100 *         sum(N_TAXA[REPRO_MA3_noT == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_nonlithophil_ExclSchool = 100 *         sum(N_TAXA[REPRO_NE == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_serialspawner_ExclSchool = 100 *         sum(N_TAXA[REPRO_SER == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_simplelithophil_ExclSchool = 100 *         sum(N_TAXA[REPRO_SILI == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_detritivore_ExclSchool = 100 *         sum(N_TAXA[TROPHIC_DE == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_gen_ExclSchool = 100 *         sum(N_TAXA[TROPHIC_GE == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_herbivore_ExclSchool = 100 *         sum(N_TAXA[TROPHIC_HB == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_insctCypr_ExclSchool = 100 *         sum(N_TAXA[TROPHIC_IN_CYP == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_insectivore_notoler_ExclSchool = 100 *         sum(N_TAXA[TROPHIC_IN_noT == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_piscivore_ExclSchool = 100 *         sum(N_TAXA[TROPHIC_PI == TRUE &amp; TYPE_SCHOOL == FALSE],             na.rm = TRUE)/ni_total_ExclSchool, pi_tv_intol_ExclSchool = 100 *         sum(N_TAXA[TOLER_I == TRUE &amp; TYPE_SCHOOL == FALSE], na.rm = TRUE)/ni_total_ExclSchool,     pi_tv_intolcoldwater_ExclSchool = 100 * sum(N_TAXA[TOLER_ICW ==         TRUE &amp; TYPE_SCHOOL == FALSE], na.rm = TRUE)/ni_total_ExclSchool,     pi_tv_sens_ExclSchool = 100 * sum(N_TAXA[TOLER_S == TRUE &amp;         TYPE_SCHOOL == FALSE], na.rm = TRUE)/ni_total_ExclSchool,     pi_tv_senscoldwater_ExclSchool = 100 * sum(N_TAXA[TOLER_SCW ==         TRUE &amp; TYPE_SCHOOL == FALSE], na.rm = TRUE)/ni_total_ExclSchool,     pi_tv_toler_ExclSchool = 100 * sum(N_TAXA[TOLER_T == TRUE &amp;         TYPE_SCHOOL == FALSE], na.rm = TRUE)/ni_total_ExclSchool,     pi_tv_tolercoldwater_ExclSchool = 100 * sum(N_TAXA[TOLER_TCW ==         TRUE &amp; TYPE_SCHOOL == FALSE], na.rm = TRUE)/ni_total_ExclSchool,     pi_exotic_ExclSchool = 100 * sum(N_TAXA[TYPE_EX == TRUE &amp;         TYPE_SCHOOL == FALSE], na.rm = TRUE)/ni_total_ExclSchool,     pi_minnow_notoler_ExclSchool = 100 * sum(N_TAXA[TYPE_MIN_noT ==         TRUE &amp; TYPE_SCHOOL == FALSE], na.rm = TRUE)/ni_total_ExclSchool,     pi_Perciformes_ExclSchool = 100 * sum(N_TAXA[TYPE_PERC ==         TRUE &amp; TYPE_SCHOOL == FALSE], na.rm = TRUE)/ni_total_ExclSchool,     pi_pioneer_ExclShool = 100 * sum(N_TAXA[TYPE_PI == TRUE &amp;         TYPE_SCHOOL == FALSE], na.rm = TRUE)/ni_total_ExclSchool,     pi_shortlived_ExclSchool = 100 * sum(N_TAXA[TYPE_SL == TRUE &amp;         TYPE_SCHOOL == FALSE], na.rm = TRUE)/ni_total_ExclSchool,     pi_dom02_ExclSchool = 100 * max(ni_dom02_ExclSchool, na.rm = TRUE)/ni_total_ExclSchool,     pt_natcoldwater = 100 * nt_natcoldwater/nt_total, pt_serialspawner = 100 *         nt_serialspawner/nt_total, pt_simplelithophil = 100 *         nt_simplelithophil/nt_total, pt_tv_sens = 100 * nt_tv_sens/nt_total,     pt_tv_senscoldwater = 100 * nt_tv_senscoldwater/nt_total,     pt_tv_toler = 100 * nt_tv_toler/nt_total, pt_tv_vtoler = 100 *         nt_tv_vtoler/nt_total, pt_beninsct_notoler = 100 * nt_beninsct_notoler/nt_total,     pt_detritivore = 100 * nt_detritivore/nt_total, pt_gen = 100 *         nt_gen/nt_total, pt_insectivore_notoler = 100 * nt_insectivore_notoler/nt_total,     pt_omnivore = 100 * nt_omnivore/nt_total, pt_darterscultpinsucker = 100 *         nt_darterscultpinsucker/nt_total, pt_pioneer = 100 *         nt_pioneer/nt_total, ni_m2_notoler = ni_total_notoler_mn/area_m2,     pi_delt_ExclSchool = 100 * sum(N_ANOMALIES[TYPE_SCHOOL !=         TRUE], na.rm = TRUE)/ni_total_ExclSchool):</span> <span style="color: #00BBBB;">ℹ</span> In argument: `x_BCG_Mean = mean(TOLVAL2, na.rm = TRUE)`.</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> In group 1: `SAMPLEID = "CHGX-432-S-1111"`, `INDEX_NAME = "MBSS_2005_Fish"`,</span>
<span class="r-err co"><span class="r-pr">#&gt;</span>   `INDEX_CLASS = "EPIEDMONT"`, `SAMP_WIDTH_M = 1.5`, `SAMP_LENGTH_M = 75`.</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span style="font-weight: bold;">Caused by error:</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> object 'TOLVAL2' not found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># View Results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html" class="external-link">View</a></span><span class="op">(</span><span class="va">df_metric_values_fish</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span></span></span>
<span class="r-in"><span><span class="co"># Example 5, periphyton (algae) metrics</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># df_metric_values_periphyton &lt;- metric.values(data_diatom_mmi_dev, "algae")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># View Results</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Erik W. Leppo.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

